manticore/core/executor.py

```diff
@@ -444,401 +382,172 @@ def generate_testcase(self, state, message = 'Testcase generated'):
+                setstate(new_state, new_value) 
+                #save the state to secondary storage
+                state_id = self.backup(new_state)
+                #add the state to the list of pending states
+                self.put(state_id)
+                #maintain a list of childres for logging purpose
+                children.append(state_id)
+        
+        logger.debug("Forking current state into states %r",children)
+        return None
 
     def run(self):
         '''
         Entry point of the Executor; called by workers to start analysis.
         '''
-        if self.profiling:
-            self._profile.enable()
 
-        policy_order=self.policy_order
-        policy=self.policy
+        #policy_order=self.policy_order
+        #policy=self.policy
 
         count = 0
         current_state = None
-        current_icount = 0
-        local_visited = set()
+        current_state_id = None
 
-        with DelayedKeyboardInterrupt():
+        with DelayedKeyboardInterrupt(self.shutdown):
             #notify siblings we are about to start a run
-            self._startRun()
+            self._start_run()
 
-        logger.debug("Starting Manticore Symbolic Emulator mainloop (pid %d).",os.getpid())
+            logger.debug("Starting Manticore Symbolic Emulator Worker (pid %d).",os.getpid())
 
 
-        while not self.isShutdown():
-            try:
-                #select a suitable state to analyze
-                if current_state is None:
-
-                    #notify siblings we are about to stop this run
-                    self._stopRun()
-
-                    with self._lock:
-                        current_state = self._getState(policy=policy, order=policy_order)
-                        assert current_state is None or current_state.constraints._child is None
-                        self._running.value+=1
-
-                    #notify siblings we are about to start a run
-                    #self._startRun()
-
-                    if current_state is None:
-                        logger.debug("No more states in the queue, byte bye!")
-                        break
-                    assert current_state is not None
-
-                    logger.setState(current_state.co)
-                    current_icount = 0
-
+            while not self.is_shutdown():
                 try:
-                    if current_state:
-                        assert solver.check(current_state.constraints)
-
-                    #execute until exception or finish or max consecutive instructions without exception 
-                    max_iters = self.dump_every + count
-
-                    # allow us to terminate manticore processes
-                    while not self.isShutdown():
-                        # Make sure current instruction is decoded so that hooks can access it
-                        current_state.cpu.decode_instruction(current_state.cpu.PC)
```

Comment:
will_execute_instruction signal now is called only when the instruction was successfully decoded. 
So _I think_ we don't need to force decoding here.